File: dbflow/src/main/java/com/raizlabs/android/dbflow/sql/language/Insert.java
package com.raizlabs.android.dbflow.sql.language;

import android.content.ContentValues;
import android.support.annotation.NonNull;

import com.raizlabs.android.dbflow.annotation.ConflictAction;
import com.raizlabs.android.dbflow.config.FlowManager;
import com.raizlabs.android.dbflow.sql.Query;
import com.raizlabs.android.dbflow.sql.QueryBuilder;
import com.raizlabs.android.dbflow.sql.language.property.IProperty;
import com.raizlabs.android.dbflow.structure.BaseModel;
import com.raizlabs.android.dbflow.structure.ModelAdapter;
import com.raizlabs.android.dbflow.structure.database.DatabaseWrapper;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.List;
import java.util.Map;

/**
 * Description: The SQLite INSERT command
 */
public class Insert<TModel> extends BaseQueriable<TModel> implements Query {


    /**
     * The columns to specify in this query (optional)
     */
    private IProperty[] columns;

    /**
     * The values to specify in this query
     */
    private List<Collection<Object>> valuesList;

    /**
     * The conflict algorithm to use to resolve inserts.
     */
    private ConflictAction conflictAction = ConflictAction.NONE;

    private From<?> selectFrom;

    /**
     * Constructs a new INSERT command
     *
     * @param table The table to insert into
     */
    public Insert(Class<TModel> table) {
        super(table);
    }

    /**
     * The optional columns to specify. If specified, the values length must correspond to these columns, and
     * each column has a 1-1 relationship to the values.
     *
     * @param columns The columns to use
     */
    @NonNull
    public Insert<TModel> columns(String... columns) {
        this.columns = new IProperty[columns.length];
        ModelAdapter<TModel> modelClassModelAdapter = FlowManager.getModelAdapter(getTable());
        for (int i = 0; i < columns.length; i++) {
            String column = columns[i];
            this.columns[i] = modelClassModelAdapter.getProperty(column);
        }
        return this;
    }

    @NonNull
    public Insert<TModel> columns(IProperty... properties) {
        this.columns = new IProperty[properties.length];
        for (int i = 0; i < properties.length; i++) {
            columns[i] = properties[i];
        }
        return this;
    }

    @NonNull
    public Insert<TModel> columns(@NonNull List<IProperty> properties) {
        return columns(properties.toArray(new IProperty[properties.size()]));
    }

    /**
     * @return Appends a list of columns to this INSERT statement from the associated {@link TModel}.
     */
    @NonNull
    public Insert<TModel> asColumns() {
        columns(FlowManager.getModelAdapter(getTable()).getAllColumnProperties());
        return this;
    }

    /**
     * @return Appends a list of columns to this INSERT and ? as the values.
     */
    @NonNull
    public Insert<TModel> asColumnValues() {
        asColumns();
        if (columns != null) {
            List<Object> values = new ArrayList<>();
            for (int i = 0; i < columns.length; i++) {
                values.add("?");
            }
            valuesList.add(values);
        }
        return this;
    }

    /**
     * The required values to specify. It must be non-empty and match the length of the columns when
     * a set of columns are specified.
     *
     * @param values The non type-converted values
     */
    @NonNull
    public Insert<TModel> values(Object... values) {
        if (this.valuesList == null) {
            this.valuesList = new ArrayList<>();
        }
        this.valuesList.add(Arrays.asList(values));
        return this;
    }

    /**
     * The required values to specify. It must be non-empty and match the length of the columns when
     * a set of columns are specified.
     *
     * @param values The non type-converted values
     */
    @NonNull
    public Insert<TModel> values(Collection<Object> values) {
        if (this.valuesList == null) {
            this.valuesList = new ArrayList<>();
        }
        this.valuesList.add(values);
        return this;
    }

    /**
     * Uses the {@link Operator} pairs to fill this insert query.
     *
     * @param conditions The conditions that we use to fill the columns and values of this INSERT
     */
    @NonNull
    public Insert<TModel> columnValues(SQLOperator... conditions) {

        String[] columns = new String[conditions.length];
        Object[] values = new Object[conditions.length];

        for (int i = 0; i < conditions.length; i++) {
            SQLOperator condition = conditions[i];
            columns[i] = condition.columnName();
            values[i] = condition.value();
        }

        return columns(columns).values(values);
    }

    /**
     * Uses the {@link Operator} pairs to fill this insert query.
     *
     * @param operatorGroup The OperatorGroup to use
     */
    @NonNull
    public Insert<TModel> columnValues(OperatorGroup operatorGroup) {

        int size = operatorGroup.size();
        String[] columns = new String[size];
        Object[] values = new Object[size];

        for (int i = 0; i < size; i++) {
            SQLOperator condition = operatorGroup.getConditions().get(i);
            columns[i] = condition.columnName();
            values[i] = condition.value();
        }

        return columns(columns).values(values);
    }

    @NonNull
    public Insert<TModel> columnValues(ContentValues contentValues) {
        java.util.Set<Map.Entry<String, Object>> entries = contentValues.valueSet();
        int count = 0;
        String[] columns = new String[contentValues.size()];
        Object[] values = new Object[contentValues.size()];
        for (Map.Entry<String, Object> entry : entries) {
            String key = entry.getKey();
            columns[count] = key;
            values[count] = contentValues.get(key);
            count++;
        }

        return columns(columns).values(values);
    }

    /**
     * Appends the specified {@link From}, which comes from a {@link Select} statement.
     *
     * @param selectFrom The from that is continuation of {@link Select}.
     */
    @NonNull
    public Insert<TModel> select(From<?> selectFrom) {
        this.selectFrom = selectFrom;
        return this;
    }


    /**
     * Specifies the optional OR method to use for this insert query
     *
     * @param action The conflict action to use
     * @return
     */
    @NonNull
    public Insert<TModel> or(ConflictAction action) {
        conflictAction = action;
        return this;
    }

    /**
     * Specifies OR REPLACE, which will either insert if row does not exist, or replace the value if it does.
     *
     * @return
     */
    @NonNull
    public Insert<TModel> orReplace() {
        return or(ConflictAction.REPLACE);
    }

    /**
     * Specifies OR ROLLBACK, which will cancel the current transaction or ABORT the current statement.
     *
     * @return
     */
    @NonNull
    public Insert<TModel> orRollback() {
        return or(ConflictAction.ROLLBACK);
    }

    /**
     * Specifies OR ABORT, which will cancel the current INSERT, but all other operations will be preserved in
     * the current transaction.
     *
     * @return
     */
    @NonNull
    public Insert<TModel> orAbort() {
        return or(ConflictAction.ABORT);
    }

    /**
     * Specifies OR FAIL, which does not back out of the previous statements. Anything else in the current
     * transaction will fail.
     *
     * @return
     */
    @NonNull
    public Insert<TModel> orFail() {
        return or(ConflictAction.FAIL);
    }

    /**
     * Specifies OR IGNORE, which ignores any kind of error and proceeds as normal.
     *
     * @return
     */
    @NonNull
    public Insert<TModel> orIgnore() {
        return or(ConflictAction.IGNORE);
    }

    @Override
    public long executeUpdateDelete(DatabaseWrapper databaseWrapper) {
        throw new IllegalStateException("Cannot call executeUpdateDelete() from an Insert");
    }

    @Override
    public long executeUpdateDelete() {
        throw new IllegalStateException("Cannot call executeUpdateDelete() from an Insert");
    }

    @Override
    public String getQuery() {
        QueryBuilder queryBuilder = new QueryBuilder("INSERT ");
        if (conflictAction != null && !conflictAction.equals(ConflictAction.NONE)) {
            queryBuilder.append("OR").appendSpaceSeparated(conflictAction);
        }
        queryBuilder.append("INTO")
            .appendSpace()
            .append(FlowManager.getTableName(getTable()));

        if (columns != null) {
            queryBuilder.append("(")
                .appendArray((Object[]) columns)
                .append(")");
        }

        // append FROM, which overrides values
        if (selectFrom != null) {
            queryBuilder.appendSpace().append(selectFrom.getQuery());
        } else {
            if (valuesList == null || valuesList.size() < 1) {
                throw new IllegalStateException("The insert of " + FlowManager.getTableName(getTable()) + " should have" +
                    "at least one value specified for the insert");
            } else if (columns != null) {
                for (Collection<Object> values : valuesList) {
                    if (values.size() != columns.length) {
                        throw new IllegalStateException("The Insert of " + FlowManager.getTableName(getTable()) + " when specifying" +
                            "columns needs to have the same amount of values and columns");
                    }
                }
            }

            queryBuilder.append(" VALUES(");
            for (int i = 0; i < valuesList.size(); i++) {
                if (i > 0) {
                    queryBuilder.append(",(");
                }
                queryBuilder.append(BaseOperator.joinArguments(", ", valuesList.get(i))).append(")");
            }
        }

        return queryBuilder.getQuery();
    }

    @Override
    public BaseModel.Action getPrimaryAction() {
        return BaseModel.Action.INSERT;
    }
}

File: dbflow/src/main/kotlin/com/raizlabs/dbflow5/query/Insert.kt
package com.raizlabs.dbflow5.query

import android.content.ContentValues
import com.raizlabs.dbflow5.annotation.ConflictAction
import com.raizlabs.dbflow5.appendArray
import com.raizlabs.dbflow5.config.FlowManager
import com.raizlabs.dbflow5.config.modelAdapter
import com.raizlabs.dbflow5.database.DatabaseWrapper
import com.raizlabs.dbflow5.query.property.IProperty
import com.raizlabs.dbflow5.sql.Query
import com.raizlabs.dbflow5.structure.ChangeAction

/**
 * Description: The SQLite INSERT command
 */
class Insert<TModel : Any>
/**
 * Constructs a new INSERT command
 *
 * @param table The table to insert into
 */
internal constructor(table: Class<TModel>)
    : BaseQueriable<TModel>(table), Query {

    /**
     * The columns to specify in this query (optional)
     */
    private var columns: List<IProperty<*>>? = null

    /**
     * The values to specify in this query
     */
    private var valuesList: MutableList<MutableList<Any?>> = arrayListOf()

    /**
     * The conflict algorithm to use to resolve inserts.
     */
    private var conflictAction: ConflictAction? = ConflictAction.NONE

    private var selectFrom: From<*>? = null

    override// append FROM, which overrides values
    val query: String
        get() {
            val queryBuilder = StringBuilder("INSERT ")
            if (conflictAction != null && conflictAction != ConflictAction.NONE) {
                queryBuilder.append("OR").append(" $conflictAction ")
            }
            queryBuilder.append("INTO")
                    .append(" ")
                    .append(FlowManager.getTableName(table))

            columns?.let { columns ->
                queryBuilder.append("(")
                        .appendArray(*columns.toTypedArray())
                        .append(")")
            }
            if (selectFrom != null) {
                queryBuilder.append(" ").append(selectFrom!!.query)
            } else {
                if (valuesList.size < 1) {
                    throw IllegalStateException("The insert of " + FlowManager.getTableName(table) + " should have" +
                            "at least one value specified for the insert")
                } else columns?.let { columns ->
                    valuesList.asSequence()
                            .filter { it.size != columns.size }
                            .forEach {
                                throw IllegalStateException(
                                        """The Insert of ${FlowManager.getTableName(table)}
                                            |when specifyingcolumns needs to have the same amount
                                            |of values and columns""".trimMargin())
                            }
                }

                queryBuilder.append(" VALUES(")
                valuesList.forEachIndexed { i, valuesList ->
                    if (i > 0) {
                        queryBuilder.append(",(")
                    }
                    queryBuilder.append(BaseOperator.joinArguments(", ", valuesList)).append(")")
                }
            }

            return queryBuilder.toString()
        }

    override val primaryAction: ChangeAction
        get() = ChangeAction.INSERT

    /**
     * The optional columns to specify. If specified, the values length must correspond to these columns, and
     * each column has a 1-1 relationship to the values.
     *
     * @param columns The columns to use
     */
    fun columns(vararg columns: String) = apply {
        val modelClassModelAdapter = table.modelAdapter
        this.columns = columns.map { modelClassModelAdapter.getProperty(it) }
    }

    fun columns(vararg properties: IProperty<*>) = apply {
        this.columns = properties.toList()
    }

    fun columns(properties: List<IProperty<*>>) = apply {
        this.columns = properties
    }

    /**
     * @return Appends a list of columns to this INSERT statement from the associated [TModel].
     */
    fun asColumns() = apply {
        columns(*table.modelAdapter.allColumnProperties)
    }

    /**
     * @return Appends a list of columns to this INSERT and ? as the values.
     */
    fun asColumnValues() = apply {
        asColumns()
        if (columns != null) {
            val values = arrayListOf<Any?>()
            for (i in columns!!.indices) {
                values.add("?")
            }
            valuesList.add(values)
        }
    }

    /**
     * The required values to specify. It must be non-empty and match the length of the columns when
     * a set of columns are specified.
     *
     * @param values The non type-converted values
     */
    fun values(vararg values: Any) = apply {
        valuesList.add(values.toMutableList())
    }

    /**
     * The required values to specify. It must be non-empty and match the length of the columns when
     * a set of columns are specified.
     *
     * @param values The non type-converted values
     */
    fun values(values: List<Any?>) = apply {
        valuesList.add(values.toMutableList())
    }

    /**
     * Uses the [Operator] pairs to fill this insert query.
     *
     * @param conditions The conditions that we use to fill the columns and values of this INSERT
     */
    fun columnValues(vararg conditions: SQLOperator) = apply {
        val columns = arrayListOf<String>()
        val values = arrayListOf<Any?>()

        for (i in conditions.indices) {
            val condition = conditions[i]
            columns += condition.columnName()
            values += condition.value()
        }

        return columns(*columns.toTypedArray()).values(values)
    }

    /**
     * Uses the [Operator] pairs to fill this insert query.
     *
     * @param operatorGroup The OperatorGroup to use
     */
    fun columnValues(operatorGroup: OperatorGroup) = apply {
        val size = operatorGroup.size
        val columns = arrayListOf<String>()
        val values = arrayListOf<Any?>()

        for (i in 0 until size) {
            val condition = operatorGroup.conditions[i]
            columns += condition.columnName()
            values += condition.value()
        }

        return columns(*columns.toTypedArray()).values(values)
    }

    fun columnValues(contentValues: ContentValues) = apply {
        val entries = contentValues.valueSet()
        val columns = arrayListOf<String>()
        val values = arrayListOf<Any?>()
        for ((key) in entries) {
            columns += key
            values += contentValues.get(key)
        }

        return columns(*columns.toTypedArray()).values(values)
    }

    /**
     * Appends the specified [From], which comes from a [Select] statement.
     *
     * @param selectFrom The from that is continuation of [Select].
     */
    fun select(selectFrom: From<*>) = apply {
        this.selectFrom = selectFrom
    }


    /**
     * Specifies the optional OR method to use for this insert query
     *
     * @param action The conflict action to use
     * @return
     */
    fun or(action: ConflictAction) = apply {
        conflictAction = action
    }

    /**
     * Specifies OR REPLACE, which will either insert if row does not exist, or replace the value if it does.
     *
     * @return
     */
    fun orReplace() = apply {
        return or(ConflictAction.REPLACE)
    }

    /**
     * Specifies OR ROLLBACK, which will cancel the current transaction or ABORT the current statement.
     *
     * @return
     */
    fun orRollback() = apply {
        return or(ConflictAction.ROLLBACK)
    }

    /**
     * Specifies OR ABORT, which will cancel the current INSERT, but all other operations will be preserved in
     * the current transaction.
     *
     * @return
     */
    fun orAbort() = apply {
        return or(ConflictAction.ABORT)
    }

    /**
     * Specifies OR FAIL, which does not back out of the previous statements. Anything else in the current
     * transaction will fail.
     *
     * @return
     */
    fun orFail() = apply {
        return or(ConflictAction.FAIL)
    }

    /**
     * Specifies OR IGNORE, which ignores any kind of error and proceeds as normal.
     *
     * @return
     */
    fun orIgnore() = apply {
        return or(ConflictAction.IGNORE)
    }

    override fun executeUpdateDelete(databaseWrapper: DatabaseWrapper): Long {
        throw IllegalStateException("Cannot call executeUpdateDelete() from an Insert")
    }
}

infix fun <T : Any> Insert<T>.orReplace(into: Array<out Pair<IProperty<*>, *>>) = orReplace().columnValues(*into)

infix fun <T : Any> Insert<T>.orRollback(into: Array<out Pair<IProperty<*>, *>>) = orRollback().columnValues(*into)

infix fun <T : Any> Insert<T>.orAbort(into: Array<out Pair<IProperty<*>, *>>) = orAbort().columnValues(*into)

infix fun <T : Any> Insert<T>.orFail(into: Array<out Pair<IProperty<*>, *>>) = orFail().columnValues(*into)

infix fun <T : Any> Insert<T>.orIgnore(into: Array<out Pair<IProperty<*>, *>>) = orIgnore().columnValues(*into)

infix fun <T : Any> Insert<T>.select(from: From<*>): Insert<T> = select(from)

fun columnValues(vararg pairs: Pair<IProperty<*>, *>): Array<out Pair<IProperty<*>, *>> = pairs

fun <T : Any> Insert<T>.columnValues(vararg pairs: Pair<IProperty<*>, *>): Insert<T> {
    val columns: MutableList<IProperty<*>> = mutableListOf()
    val values = mutableListOf<Any?>()
    pairs.forEach {
        columns.add(it.first)
        values.add(it.second)
    }
    this.columns(columns).values(values)
    return this
}
